datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java

```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
       return _zkClient;
     }
 
+    public class ZkClientMockStateChangeListener extends ZkStateChangeListener {
+      boolean sessionExpired = false;
+      @Override
+      public void handleStateChanged(Watcher.Event.KeeperState state) {
+        super.handleStateChanged(state);
+        if (state == Watcher.Event.KeeperState.Expired) {
+          LOG.info("ZkStateChangeListener::Session expired.");
+          sessionExpired = true;
+        }
+      }
+    }
+
+    @Override
+    ZkStateChangeListener getOrCreateStateChangeListener() {
+      _zkClientMockStateChangeListener = spy(new ZkClientMockStateChangeListener());
+      return _zkClientMockStateChangeListener;
+    }
+
     public ZkClient getZkClient() {
       return _zkClient;
     }
+
+    public ZkClientMockStateChangeListener getZkStateChangeListener() {
+      return _zkClientMockStateChangeListener;
+    }
+  }
+
+  @Test
+  public void testZookeeperSessionExpiry() throws InterruptedException {
+    String testCluster = "testDeleteTaskWithPrefix";
+    String connectorType = "connectorType";
+    Duration timeout = Duration.ofMinutes(1);
+
+    ZkClientInterceptingAdapter adapter = createInterceptingZkAdapter(testCluster, 5000);
+    adapter.connect();
+
+    DatastreamTaskImpl task = new DatastreamTaskImpl();
+    task.setId("3");
+    task.setConnectorType(connectorType);
+    task.setZkAdapter(adapter);
+
+    List<DatastreamTask> tasks = new ArrayList<>();
+    tasks.add(task);
+    updateInstanceAssignment(adapter, adapter.getInstanceName(), tasks);
+
+    LOG.info("Acquire from instance1 should succeed");
+    Assert.assertTrue(expectException(() -> task.acquire(timeout), false));
+
+    simulateSessionExpiration(adapter);
+
+    Thread.sleep(5000);
+    Assert.assertTrue(adapter.getZkStateChangeListener().sessionExpired);
+    Mockito.verify(adapter, Mockito.times(1)).onSessionExpired();
+  }
+
+  private void simulateSessionExpiration(ZkClientInterceptingAdapter adapter) {
+    ZkConnection zkConnection = null;
+    try {
+      Field privateField = ZkClient.class.getSuperclass().getDeclaredField("_connection");
+      privateField.setAccessible(true);
+      zkConnection = (ZkConnection) privateField.get(adapter.getZkClient());
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      Assert.fail(e.toString());
+    }
+
+    ZooKeeper zookeeper = zkConnection.getZookeeper();
+    long sessionId = zookeeper.getSessionId();
+
+    LOG.info("Closing/expiring session:" + sessionId);
+    ZooKeeperServer zkServer = _embeddedZookeeper.getZooKeeperServer();
+    zkServer.closeSession(sessionId);
```

Comment:
Unless you foresee a good reason to keep it the way it is, I'd suggest exposing a `closeSession()` on `EmbeddedZookeeper` to limit the exposed internals.